### PR TITLE
dotCMS/core#22164 Whitelist blocks 

### LINKED ISSUE #22164: Whitelist blocks
**Is your feature request related to a problem? Please describe.**
As a user, I want to be able to limit the blocks a content creator can add to the block editor field.

For example, some users might don't want to allow `<h1>` because there is one already in the page.

**Describe the solution you'd like**
Users needs to be able to add a field variable with the list of blocks they want to show.

Consider that for headings, we need to be granular with the levels.

**Describe alternatives you've considered**
N/A

**Additional context**
Consider there are other field variables, for this one I think we can use `allowedBlocks`

We can't remove the blocks just from the array list because if the block exist and is init in the editor users will be able to copy and paste blocks, like, let's say I don't want an `h1` and I don't put it in the list, still user will be able to paste a `h1` from copy and paste content.

We need to import blocks individually, not using the `StarterKit`.

We need to [lazy load the editor](https://github.com/dotCMS/core/issues/22171#issuecomment-1119923323).

### PR DESCRIPTION
**Pull request by hmoreras**
### Proposed Changes
* Allow the users to set a restriction of blocks in the block editor. 
* For details check: https://github.com/dotcms/core/issues/22164#issuecomment-1179319411

### CODE REVIEW

File: core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts [hunk 1]
```diff
@@ -42,42 +45,91 @@ export class DotBlockEditorComponent implements OnInit {
...
+                    case 'codeBlock':
+                        delete starterKitOptions.codeBlock;
+                        break;
+                    case 'horizontalRule':
+                        delete starterKitOptions.horizontalRule;
+                        break;
+                    case 'contentlets':
+                        customExtension.push(ContentletBlock(this.injector));
+                }
+            });
```
Comment (fmontes) [inline core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts:124]: What about something like this:

```
let starterKitOptions = {
  heading: false,
  orderedList: false,
  bulletList: false,
  blockquote: false,
  codeBlock: false,
  horizontalRule: false
};

const whiteList = ['orderedList', 'blockquote', 'heading1'];

const headings = whiteList.filter(item => item.startsWith('heading')).length

whiteList.forEach(item => {
  if (starterKitOptions[item]) {
    starterKitOptions = {
      ...starterKitOptions,
      [item]: true
    }
  }
})
```

File: core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts [hunk 2]
```diff
@@ -42,42 +45,91 @@ export class DotBlockEditorComponent implements OnInit {
...
+            starterKitOptions.heading = headingOptions.levels.length ? headingOptions : false;
```
Comment (fmontes) [inline core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts:125]: What about something like this?

```
const whiteList = ['orderedList', 'blockquote', 'heading1'];
const headings = whiteList.filter(item => item.startsWith('heading')).length
starterKitOptions.heading = headings || false
```

File: core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts [hunk 3]
```diff
@@ -42,42 +45,91 @@ export class DotBlockEditorComponent implements OnInit {
...
-                        if (node.type.name === 'heading') {
-                            return `${toTitleCase(node.type.name)} ${node.attrs.level}`;
-                        }
-
-                        return 'Type "/" for commmands';
+        const defaultExtensions: Extensions = [
+            DotConfigExtension({
+                lang: this.lang,
+                allowedContentTypes: this.allowedContentTypes,
+                allowedBlocks: this.setAllowedBlocks()
```
Comment (fmontes) [inline core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts:52]: You can do what happens in the `setAllowedBlocks` if you create a setter in the `Input()`

File: core-web/libs/block-editor/src/lib/plugins/dot-bubble-menu.plugin.ts [hunk 4]
```diff
@@ -308,7 +308,12 @@ export class DotBubbleMenuPluginView extends BubbleMenuView {
     }
 
     setChangeToOptions() {
-        const changeToOptions = suggestionOptions.filter((item) => item.id != 'horizontalLine');
+        const allowedBlocks: string[] = this.editor.storage.dotConfig.allowedBlocks;
+
+        const changeToOptions =
+            allowedBlocks.length > 1
+                ? suggestionOptions.filter((item) => allowedBlocks.includes(item.id))
+                : suggestionOptions.filter((item) => item.id != 'horizontalLine');
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/plugins/dot-bubble-menu.plugin.ts:316]: What happens if I set a block in the white list and is misspelled?

Comment (fmontes) [reply core-web/libs/block-editor/src/lib/plugins/dot-bubble-menu.plugin.ts:316]: What about passing this from the init? I think that is better is we can pass the block from the init and plugins and extension are only in charge of render what they get, a more centralized way.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/block-editor/src/lib/plugins/dot-bubble-menu.plugin.ts (hunk 4 above)
Comment (hmoreras) [reply core-web/libs/block-editor/src/lib/plugins/dot-bubble-menu.plugin.ts:316]: if is misspelled we ignore it. About the centralized way,  the "change" to will have a limited blocks, and will not be the same a suggestions. for that propuse we hace a "core" list `suggestionOptions`

**Review by rjvelazco — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts [hunk 5]
```diff
@@ -35,49 +37,89 @@ export class DotBlockEditorComponent implements OnInit {
...
+            }
+        });
+
+        return {
+            heading: headingOptions.levels.length ? headingOptions : false,
+            ...(this._allowedBlocks['orderedList'] ? {} : {orderedList: false}),
+            ...(this._allowedBlocks['bulletList'] ? {} : {bulletList: false}),
+            ...(this._allowedBlocks['blockquote'] ? {} : {blockquote: false}),
+            ...(this._allowedBlocks['codeBlock'] ? {} : {codeBlock: false}),
+            ...(this._allowedBlocks['horizontalRule'] ? {} : {horizontalRule: false})
```
Comment (rjvelazco) [inline core-web/apps/dotcms-block-editor/src/app/dot-block-editor/dot-block-editor.component.ts:114]: I think we can do something like:

```
// This convert the array of allow Items into a object of type: { 'item': true }
const staterKitOptions = this._allowedBlocks.reduce((object, item) => ({ ...object, [item]: true}), {});
return {
      heading: headingOptions.levels.length ? headingOptions : false,
       ...staterKitOptions
};
```

**Review by rjvelazco — CHANGES_REQUESTED** [review]

**Review by rjvelazco — APPROVED** [review]

**Review by fmontes — CHANGES_REQUESTED** [review]

**Pull request merged**